File: TabCommon/Library/src/main/java/com/nextzy/library/base/mvvm/layer1View/BaseMvvmIndividualListAdapter.java
package com.nextzy.library.base.mvvm.layer1View;


import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.RecyclerView;

import com.nextzy.library.base.mvvm.layer2ViewModel.BaseListAdapterViewModel;
import com.nextzy.library.base.view.holder.base.BaseViewHolder;

import java.util.List;


/**
 * Created by「 The Khaeng 」on 20 Aug 2017 :)
 */

public abstract class BaseMvvmIndividualListAdapter<VH extends BaseViewHolder, VM extends BaseListAdapterViewModel>
        extends BaseMvvmListAdapter<VH,VM>{


    private RecyclerView recyclerView;

    public BaseMvvmIndividualListAdapter( FragmentActivity activity ){
        super(activity);
    }

    public BaseMvvmIndividualListAdapter( Fragment fragment ){
        super(fragment);
    }

    public abstract boolean setSharedViewModel();

    /* =================================== Read Item =============================================*/
    public List<Object> getItemList(){
        return getPrivateViewModel().getItemList();
    }

    public Object getItem( int pos ){
        return getPrivateViewModel().getItem( pos );
    }

    public boolean hasItems(){
        return !getPrivateViewModel().isEmpty();
    }

    /* =================================== Update Item ===========================================*/
    public void setItemList( List<Object> itemList ){
        setItemList( itemList, true );
    }

    public void setItemList( List<Object> itemList, boolean isNotify ){
        getPrivateViewModel().setItemList(itemList);
        if( isNotify ){
            notifyDataSetChanged();
        }
    }

    public void addItem( Object item ){
        addItem( item, true );
    }

    public void addItem( Object item, boolean isNotify ){
        getPrivateViewModel().addItem( item );
        if( isNotify ){
            notifyItemInserted( getItemCount() - 1 );
        }
    }

    public void addItem( int index, Object item ){
        addItem( index, item, true );
    }

    public void addItem( int index, Object item, boolean isNotify ){
        getPrivateViewModel().addItem( index, item );
        if( isNotify ){
            notifyItemInserted( index );
        }
    }

    public void addAllItems( List<Object> items ){
        addAllItems( items, true );
    }

    public void addAllItems( List<Object> items, boolean isNotify ){
        getPrivateViewModel().addAllItems( items );
        if( isNotify ){
            notifyItemRangeInserted( getItemCount() - 1, items.size() );
        }
    }

    public void addAllItems( int index, List<Object> items ){
        addAllItems( index, items, true );
    }

    public void addAllItems( int index, List<Object> items, boolean isNotify ){
        getPrivateViewModel().addAllItems( index, items );
        if( isNotify ){
            notifyItemRangeInserted( index, items.size() );
        }
    }

    public void updateItem( int index, Object item ){
        updateItem( index, item, true );
    }

    public void updateItem( int index, Object item, boolean isNotify ){
        getPrivateViewModel().updateItem( index, item );
        if( isNotify ){
            notifyItemChanged( index );
        }
    }

    /* =================================== Delete Item ===========================================*/
    public void removeAllItem(){
        removeAllItem( true );
    }

    public void removeAllItem( boolean isNotify ){
        getPrivateViewModel().removeAllItem();
        if( isNotify ){
            notifyDataSetChanged();
        }
    }

    public void removeItemAt( int index ){
        removeItemAt( index, true );
    }

    public void removeItemAt( int index, boolean isNotify ){
        getPrivateViewModel().removeItemAt(index);
        if( isNotify ){
            notifyItemRemoved( index );
        }
    }

    public int removeItem( Object baseItem ){
        return removeItem( baseItem, true );
    }

    public int removeItem( Object baseItem, boolean isNotify ){
        int removeIndex = -1;
        for( int i = 0; i < getItemCount(); i++ ){
            if( getItem( i ).equals( baseItem ) ){
                removeIndex = i;
                break;
            }
        }
        if( removeIndex != -1 ){
            getPrivateViewModel().removeItem( baseItem );
        }
        if( isNotify && removeIndex != -1 ){
            notifyItemRemoved( removeIndex );
        }
        return removeIndex;
    }

    @Override
    public void onBindViewHolder( VH holder, int position ){
        super.onBindViewHolder( holder,position );
        holder.setItem( getItem( holder.getAdapterPosition() ) );
    }

    private VM getPrivateViewModel(){
        if( setSharedViewModel() ){
            return getViewModelShared();
        }else{
            return getViewModel();
        }
    }

}

File: TabCommon/Library/src/main/java/com/nextzy/library/base/mvvm/layer1View/BaseMvvmIndividualListAdapter.kt
package com.nextzy.library.base.mvvm.layer1View


import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import com.nextzy.library.base.mvvm.layer2ViewModel.BaseListAdapterViewModel
import com.nextzy.library.base.view.holder.base.BaseViewHolder


/**
 * Created by「 The Khaeng 」on 20 Aug 2017 :)
 */

abstract class BaseMvvmIndividualListAdapter<VH : BaseViewHolder<Any>, VM : BaseListAdapterViewModel>
    : BaseMvvmListAdapter<VH, VM> {


    var itemList: MutableList<Any>
        get() = privateViewModel?.getItemList() ?: mutableListOf()
        set(itemList) = setItemList(itemList, true)

    private val privateViewModel: VM?
        get() = if (setSharedViewModel()) {
            viewModelShared
        } else {
            viewModel
        }

    constructor(activity: FragmentActivity) : super(activity) {}

    constructor(fragment: Fragment) : super(fragment) {}


    /* =================================== Read Item =============================================*/
    fun getItem(pos: Int): Any? {
        return privateViewModel?.getItem(pos)
    }

    fun hasItems(): Boolean {
        return privateViewModel?.isEmpty?.not() ?: false
    }

    /* =================================== Update Item ===========================================*/
    fun setItemList(itemList: MutableList<Any>, isNotify: Boolean) {
        privateViewModel?.setItemList(itemList)
        if (isNotify) {
            notifyDataSetChanged()
        }
    }

    fun addItem(index: Int = -1, item: Any, isNotify: Boolean = true) {
        if (index == -1) privateViewModel?.addItem(item)
        else privateViewModel?.addItem(index, item)
        if (isNotify) {
            notifyItemInserted(index)
        }
    }

    fun addAllItems(index: Int = -1, items: List<Any>, isNotify: Boolean = true) {
        if (index == -1) privateViewModel?.addAllItems(items)
        else privateViewModel?.addAllItems(index, items)

        if (isNotify) {
            notifyItemRangeInserted(index, items.size)
        }
    }

    fun updateItem(index: Int, item: Any, isNotify: Boolean = true) {
        privateViewModel?.updateItem(index, item)
        if (isNotify) {
            notifyItemChanged(index)
        }
    }

    /* =================================== Delete Item ===========================================*/

    fun removeAllItem(isNotify: Boolean = true) {
        privateViewModel?.removeAllItem()
        if (isNotify) {
            notifyDataSetChanged()
        }
    }

    fun removeItemAt(index: Int, isNotify: Boolean = true) {
        privateViewModel?.removeItemAt(index)
        if (isNotify) {
            notifyItemRemoved(index)
        }
    }

    fun removeItem(baseItem: Any, isNotify: Boolean = true): Int {
        val removeIndex = (0 until itemCount).firstOrNull { getItem(it) == baseItem } ?: -1
        if (removeIndex != -1) {
            privateViewModel?.removeItem(baseItem)
        }
        if (isNotify && removeIndex != -1) {
            notifyItemRemoved(removeIndex)
        }
        return removeIndex
    }

    override
    fun onBindViewHolder(holder: VH, position: Int) {
        super.onBindViewHolder(holder, position)
        holder.item = getItem(holder.adapterPosition)
    }

    abstract fun setSharedViewModel(): Boolean

}
